include/llama/Copy.hpp

```diff
@@ -0,0 +1,381 @@
+                            {
+                                for (std::size_t j = 0; j < LanesSrc; j += L)
+                                    copyLBlock(threadSrc, i + j, coord);
+                            });
+                }
+                else
+                {
+                    forEachLeaf<RecordDim>(
+                        [&](auto coord) LLAMA_LAMBDA_INLINE
+                        {
+                            auto* threadSrc = mapSrc(start, coord);
+                            for (std::size_t i = start; i < stop; i += L)
+                                copyLBlock(threadSrc, i, coord);
+                        });
+                }
+            }
+        }
+        else
+        {
+            // optimized for linear writing
+            constexpr auto dstL = dstIsAoSoA ? LanesDst : L;
+            const auto elementsPerThread = flatSize / dstL / threadCount * dstL;
+            {
+                const auto start = threadId * elementsPerThread;
+                const auto stop = threadId == threadCount - 1 ? flatSize : (threadId + 1) * elementsPerThread;
+
+                auto copyLBlock = [&](std::byte*& threadDst, std::size_t srcIndex, auto coord) LLAMA_LAMBDA_INLINE
+                {
+                    constexpr auto bytes = L * sizeof(GetType<RecordDim, decltype(coord)>);
+                    std::memcpy(threadDst, mapSrc(srcIndex, coord), bytes);
+                    threadDst += bytes;
+                };
+                if constexpr (dstIsAoSoA)
+                {
+                    auto* threadDst = mapDst(start, RecordCoord<>{});
+                    for (std::size_t i = start; i < stop; i += LanesDst)
+                        forEachLeaf<RecordDim>(
+                            [&](auto coord) LLAMA_LAMBDA_INLINE
+                            {
+                                for (std::size_t j = 0; j < LanesDst; j += L)
+                                    copyLBlock(threadDst, i + j, coord);
+                            });
+                }
+                else
+                {
+                    forEachLeaf<RecordDim>(
+                        [&](auto coord) LLAMA_LAMBDA_INLINE
+                        {
+                            auto* threadDst = mapDst(start, coord);
+                            for (std::size_t i = start; i < stop; i += L)
+                                copyLBlock(threadDst, i, coord);
+                        });
+                }
+            }
+        }
+    }
+
+    /// @brief Generic implementation of \ref copy defaulting to \ref fieldWiseCopy. LLAMA provides several
+    /// specializations of this construct for specific mappings. Users are encouraged to also specialize this template
+    /// with better copy algorithms for further combinations of mappings, if they can and want to provide a better
+    /// implementation.
+    template <typename SrcMapping, typename DstMapping, typename SFINAE = void>
+    struct Copy
```

Comment:
Would it make sense to hide these `Copy` structs in a separate namespace? Otherwise the user might be inclined to use them directly instead of the `copy()` function.

Reply:
Not sure. The user should call `copy()`. But when the user adds specializations, the need to specialize `Copy<...>`. So it should be inside a namespace that does not look scary. I think I prefer leaving it just in `llama`. But if you have a good suggestion, I might take it :)

Reply:
In alpaka we put these types into the the `alpaka::traits` namespace. Maybe that is an option here?

Reply:
I know, but `trait` is the wrong term. `Copy` is a customization point, which is pretty long to spell.

Reply:
Thinking again about this, I would like to keep it as is for now. Is that fine for you?

Reply:
Sure, LLAMA is your pet ;-)